tensorflow/contrib/opt/python/training/irprop_plus.py

```diff
@@ -0,0 +1,284 @@
+
+    ```
+    old_grad <- 0 (Initialize the gradient from the previous iteration g{t-1})
+    delta_update <- delta_zero (Initialize step-size)
+    t <- 0 (Initialize iteration counter)
+    ```
+
+    The following update rule is performed for each individual objective
+    variable (e.g., weight), where `g{t}` denotes the partial derivative of the
+    objective function with respect to the objective variable at iteration `t`:
+
+    ```
+    t <- t + 1
+    grad_sign <- sign(g{t} * g{t-1})
+    if (grad_sign > 0)
+      delta_update{t} <- min(eta_plus * delta_update{t-1}, delta_max)
+      variable{t+1} <- variable{t} -sign(g{t}) * delta_update{t}
+    else if (grad_sign < 0)
+      delta_update{t} <- max(eta_minus * delta_update{t-1}, delta_min)
+      if (error{t} > error{t-1})
+        variable{t+1} <- variable{t-1}
+      g{t} = 0
+    else
+      variable{t+1} <- variable{t} - sign(g{t}) * delta_update{t}
+
+    ```
+
+    Args:
+      eta_minus: Step-size decrease factor.
+      eta_plus: Step-size increase factor.
+      delta_zero: Initial step-size.
+      delta_min: Lower bound on step-size.
+      delta_max: Upper bound on step-size.
+      use_locking: If True, use locks for update operations.
+      name: Optional name for the operations created when applying gradients.
+         Defaults to "IRpropPlusOptimizer".
+   """
+
+    super(IRpropPlusOptimizer, self).__init__(use_locking, name)
+
+    # Init parameters
+    self._set_hyper("eta_minus", eta_minus)
+    self._set_hyper("eta_plus", eta_plus)
+    self._set_hyper("delta_zero", delta_zero)
+    self._set_hyper("delta_min", delta_min)
+    self._set_hyper("delta_max", delta_max)
+
+    # Error auxiliary var
+    self._error = None
+
+  def _create_vars(self, var_list, state):
+    # Error tensors
+    state.create_non_slot(initial_value=lambda: 0.0, name="error")
+    state.create_non_slot(initial_value=lambda: 0.0, name="old_error")
+
+    for v in var_list:
+      state.zeros_slot(v, "old_grad")
+
+      init_step = math_ops.add(
+          array_ops.zeros_like(v),
+          state.get_hyper("delta_zero", v.dtype.base_dtype))
+      state.create_slot_with_initializer(v, init_step, v.get_shape(),
+                                         v.dtype.base_dtype, "delta_update")
+
+  def _get_error_values(self, state=None):
+    if state is None:
+      state = self._get_per_graph_state()
+    return (state.get_non_slot("error"),
+            state.get_non_slot("old_error"))
+
+  # Helper method to check if variable is scalar
+  def _is_scalar(self, tensor):
+    return tensor is not None and \
+            tensor.shape.ndims == 0
+
+  def _apply_dense(self, grad, var, state):
+    old_grad = state.get_slot(var, "old_grad")
+    delta_update = state.get_slot(var, "delta_update")
+
+    error, old_error = self._get_error_values(state)
+    # Update the error E(t) passed in apply_gradients or minimize
+    update_error = error.assign(self._error, use_locking=self._use_locking)
```

Comment:
This update is getting dropped on the floor because nothing depends on it. Add a control dependency to it to make sure it executes.

Reply:
ok